eureka-client/src/main/java/com/netflix/discovery/shared/resolver/ReloadingClusterResolver.java

```diff
@@ -0,0 +1,141 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.netflix.discovery.EurekaClientNames.METRIC_RESOLVER_PREFIX;
+
+/**
+ * A cluster resolver implementation that periodically creates a new {@link ClusterResolver} instance that
+ * swaps the previous value. If the new resolver cannot be created or contains empty server list, the previous
+ * one is used. Failed requests are retried using exponential back-off strategy.
+ * <br/>
+ * It provides more insight in form of additional logging and metrics, as it is supposed to be used as a top
+ * level resolver.
+ *
+ * @author Tomasz Bak
+ */
+public class ReloadingClusterResolver implements ClusterResolver {
+
+    private static final Logger logger = LoggerFactory.getLogger(ReloadingClusterResolver.class);
+
+    private static final long MAX_RELOAD_INTERVAL_MULTIPLIER = 5;
+
+    private final ClusterResolverFactory factory;
+    private final long reloadIntervalMs;
+    private final long maxReloadIntervalMs;
+
+    private final AtomicReference<ClusterResolver> delegateRef;
+
+    private volatile long lastUpdateTime;
+    private volatile long currentReloadIntervalMs;
+
+    // Metric timestamp, tracking last time when data were effectively changed.
+    private volatile long lastReloadTimestamp = -1;
+
+    public ReloadingClusterResolver(final ClusterResolverFactory factory, final long reloadIntervalMs) {
+        this.factory = factory;
+        this.reloadIntervalMs = reloadIntervalMs;
+        this.maxReloadIntervalMs = MAX_RELOAD_INTERVAL_MULTIPLIER * reloadIntervalMs;
+        this.delegateRef = new AtomicReference<>(factory.createClusterResolver());
+
+        this.lastUpdateTime = System.currentTimeMillis();
+        this.currentReloadIntervalMs = reloadIntervalMs;
+
+        List<EurekaEndpoint> clusterEndpoints = delegateRef.get().getClusterEndpoints();
+        if (clusterEndpoints.isEmpty()) {
+            logger.error("Empty Eureka server endpoint list during initialization process");
+            throw new ClusterResolverException("Resolved to an empty endpoint list");
+        }
+
+        if (logger.isInfoEnabled()) {
+            logger.info("Initiated with delegate resolver of type {}; next reload in {}[sec]. Loaded endpoints={}",
+                    delegateRef.get().getClass(), currentReloadIntervalMs / 1000, clusterEndpoints);
+        }
+
+        try {
+            Monitors.registerObject(this);
+        } catch (Throwable e) {
+            logger.warn("Cannot register metrics", e);
+        }
+    }
+
+    @Override
+    public String getRegion() {
+        ClusterResolver delegate = delegateRef.get();
+        return delegate == null ? null : delegate.getRegion();
+    }
+
+    @Override
+    public List<EurekaEndpoint> getClusterEndpoints() {
+        long expiryTime = lastUpdateTime + currentReloadIntervalMs;
+        if (expiryTime <= System.currentTimeMillis()) {
+            try {
+                ClusterResolver newDelegate = reload();
```

Comment:
Yes. Certain resolvers take time to get their data (DNS), so this could block, and this would be in the request path. In the first iteration I have implemented reloading resolver with a background thread. Probably I should get it back.